### DISCUSSION TITLE
Redirect from action with data

### DISCUSSION THREAD

**Discussion by Falci** [remix-run/react-router#10936] [Q&A]
Use case: 
The user filled the sign up form, 
The `action` processed the request and we sent a validation email.
Next we need to redirect the user to a friendly page to let them know about the validation email.

How can we, from inside an action, redirect to the validate page, with some data from the POST request (name, email).

I was looking for something like `navigate` with state and avoid query/path params

**Comment by florianbepunkt** [comment]
Same here.

My use case: During login, some users are required to setup an MFA device. From my auth service I get a response if the login was either successful or if MFA setup is needed (along with the MFA secret for setup). Based on this I would like to redirect the user to the main page on login success or to an mfa-setup page if needed. However I need to send the MFA setup data along with the redirect.

A redirect should be basically a 302 response. I can attach data to this response like in the following example, but I haven't found a way to read this data on the new route.

```ts
  return new Response(JSON.stringify({ foo: "bar" }), {
    status: 302,
    headers: {
      "Content-Type": "application/json; utf-8",
      Location: "/mfa-setup",
    },
  });
```

So my feature request is to add a way to read data sent with a redirect on the redirected page.

**Comment by AYastrebtsov** [comment]
Same here.

Right now I just return data from action and use navigate() in useEffect.

Looks very dirty, but until redirect can accept state, I guess this is the only way

**Comment by justin-hackin** [comment]
I'm also wishing for a feature like this. I have an issue where I'm trying to perform actions in react-admin custom routes and I have to target the root admin route but then I can't get action errors in the admin subroutes so I was hoping I could redirect to the subroutes with the action data.

**Comment by mateusvahl** [comment]
Same here:

1. We have a route: `/example?ids=123,456`
2. A `useFetch()` call triggers an action that returns a `redirect('/example?ids=789,101')`.
3. We want to display a confirmation message based on the fetch response.

**The Problem:**
- We can't return both a JSON response and a redirect.
- We can't return a json & navigate with a useEffect because as soon the action is trigger, the previous ids no longer exist in the system.

Yes, we can redirect to `redirect('/example?ids=789,101&status=success')`  but not my favorite solution.

**Reply by mateusvahl** [reply]
I've found that I can programmatically skip revalidation from a particular action, example:

1 ) Your fetcher triggers an submit to `'/my-route/action'`
2 ) You can skip that revalidation based on `request.formAction`
3 ) You action no longer returns a redirect, and you can return data
4) You need manually redirect in the front-end.

```tsx
           shouldRevalidate: request => {
                  if (request.formAction === '/my-route/action') {
                    return false;
                  }
                  return true;
                },
```
